src/Analyzer.Cli/TemplateDiscovery.cs

```diff
@@ -0,0 +1,163 @@
+                Path.GetFileNameWithoutExtension(template.Name) + ".parameters*.json",
+                new EnumerationOptions
+                {
+                    MatchCasing = MatchCasing.CaseInsensitive,
+                    RecurseSubdirectories = false
+                });
+
+            return parameterFiles.Any()
+                ? parameterFiles.Select(parameters => new TemplateAndParams(template, parameters))
+                : new[] { new TemplateAndParams(template, null) };
+        }
+
+        /// <summary>
+        /// Determines whether or not a file is valid template.
+        /// If the file extension is ".bicep", it is assumed to be valid.
+        /// Otherwise, the file must be a JSON file with the proper schema for ARM templates.
+        /// </summary>
+        /// <param name="file">The file to determine template validity on.</param>
+        /// <returns>true if the file is a valid ARM template; false otherwise.</returns>
+        public static bool IsValidTemplate(FileInfo file)
+        {
+            // assume bicep files are valid, they are compiled/verified later
+            if (file.Extension.Equals(".bicep", StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+
+            using var fileStream = new StreamReader(file.OpenRead());
+            var reader = new JsonTextReader(fileStream);
+
+            reader.Read();
+            if (reader.TokenType != JsonToken.StartObject)
+            {
+                return false;
+            }
+
+            while (reader.Read())
+            {
+                if (reader.Depth == 1 && reader.TokenType == JsonToken.PropertyName)
+                {
+                    if (string.Equals((string)reader.Value, "$schema", StringComparison.OrdinalIgnoreCase))
+                    {
+                        reader.Read();
+                        if (reader.TokenType != JsonToken.String)
+                        {
+                            return false;
+                        }
+
+                        return validSchemaRegex.IsMatch((string)reader.Value);
+                    }
+                    else if (!validTemplateProperties.Any(property => string.Equals((string)reader.Value, property, StringComparison.OrdinalIgnoreCase)))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Returns the contents of the given template and parameters files.
+        /// </summary>
+        /// <param name="templateAndParams"><see cref="TemplateAndParams"/> containing the template and parameters to get the contents of.</param>
+        /// <returns>The contents of the template and parameters.</returns>
+        public static (string, string) GetTemplateAndParameterContents(TemplateAndParams templateAndParams)
+        {
+            // Logic is simple now, may expand in the future.
```

Comment:
Want to clarify more for new owners/create an issue?

Reply:
I don't have anything specific in mind, just wanted to consolidate template handling logic.

Reply:
then delete the comment